Realm/ObjectStore/object_store.cpp

```diff
@@ -0,0 +1,418 @@
+
+const uint64_t ObjectStore::NotVersioned = numeric_limits<uint64_t>::max();
+
+bool ObjectStore::has_metadata_tables(realm::Group *group) {
+    return group->get_table(c_primaryKeyTableName) && group->get_table(c_metadataTableName);
+}
+
+bool ObjectStore::create_metadata_tables(realm::Group *group) {
+    bool changed = false;
+    realm::TableRef table = group->get_or_add_table(c_primaryKeyTableName);
+    if (table->get_column_count() == 0) {
+        table->add_column(realm::type_String, c_primaryKeyObjectClassColumnName);
+        table->add_column(realm::type_String, c_primaryKeyPropertyNameColumnName);
+        changed = true;
+    }
+
+    table = group->get_or_add_table(c_metadataTableName);
+    if (table->get_column_count() == 0) {
+        table->add_column(realm::type_Int, c_versionColumnName);
+
+        // set initial version
+        table->add_empty_row();
+        table->set_int(c_versionColumnIndex, 0, realm::ObjectStore::NotVersioned);
+        changed = true;
+    }
+
+    return changed;
+}
+
+uint64_t ObjectStore::get_schema_version(realm::Group *group) {
+    realm::TableRef table = group->get_table(c_metadataTableName);
+    if (!table || table->get_column_count() == 0) {
+        return realm::ObjectStore::NotVersioned;
+    }
+    return table->get_int(c_versionColumnIndex, 0);
+}
+
+void ObjectStore::set_schema_version(realm::Group *group, uint64_t version) {
+    realm::TableRef table = group->get_or_add_table(c_metadataTableName);
+    table->set_int(c_versionColumnIndex, 0, version);
+}
+
+StringData ObjectStore::get_primary_key_for_object(realm::Group *group, StringData object_type) {
+    realm::TableRef table = group->get_table(c_primaryKeyTableName);
+    if (!table) {
+        return "";
+    }
+    size_t row = table->find_first_string(c_primaryKeyObjectClassColumnIndex, object_type);
+    if (row == realm::not_found) {
+        return "";
+    }
+    return table->get_string(c_primaryKeyPropertyNameColumnIndex, row);
+}
+
+void ObjectStore::set_primary_key_for_object(realm::Group *group, StringData object_type, StringData primary_key) {
+    realm::TableRef table = group->get_table(c_primaryKeyTableName);
+
+    // get row or create if new object and populate
+    size_t row = table->find_first_string(c_primaryKeyObjectClassColumnIndex, object_type);
+    if (row == realm::not_found && primary_key.size()) {
+        row = table->add_empty_row();
+        table->set_string(c_primaryKeyObjectClassColumnIndex, row, object_type);
+    }
+
+    // set if changing, or remove if setting to nil
+    if (primary_key.size() == 0) {
+        if (row != realm::not_found) {
+            table->remove(row);
+        }
+    }
+    else {
+        table->set_string(c_primaryKeyPropertyNameColumnIndex, row, primary_key);
+    }
+}
+
+string ObjectStore::object_type_for_table_name(const string &table_name) {
+    if (table_name.size() >= 6 && table_name.compare(0, 6, c_object_table_name_prefix) == 0) {
```

Comment:
Wouldn't it be safer to declare a constant to be the length of `c_object_table_name_prefix` - in case we change it.